server/src/main/java/org/elasticsearch/index/mapper/PassThroughObjectMapper.java

```diff
@@ -153,6 +204,59 @@ protected static void parsePassthrough(String name, Map<String, Object> node, Pa
                 );
                 node.remove(TimeSeriesParams.TIME_SERIES_DIMENSION_PARAM);
             }
+            fieldNode = node.get(SUPERSEDED_BY_PARAM);
+            if (fieldNode != null) {
+                builder.supersededBy = new TreeSet<>(Arrays.asList(nodeStringArrayValue(fieldNode)));
+                node.remove(SUPERSEDED_BY_PARAM);
+            }
+        }
+    }
+
+    /**
+     * Checks the passed objects for circular dependencies on the superseded_by cross-dependencies.
+     * @param passThroughMappers objects to check
+     * @return A list containing the circular dependency chain, or an empty list if no circular dependency is found
+     */
+    public static List<PassThroughObjectMapper> checkSupersedesForCircularDeps(Collection<PassThroughObjectMapper> passThroughMappers) {
+        if (passThroughMappers.size() < 2) {
+            return List.of();
+        }
+
+        // Perform DFS.
+        // The implementation below assumes that the graph is rather small and sparse. Further optimizations are
+        // required (e.g. use sets for lookups) if production systems end up with hundreds of tightly-coupled objects.
+        int i = 0;
```

Comment:
Could this be simplified to:

```
Map<String, PassThroughObjectMapper> lookupByName = passThroughMappers.stream()
            .collect(Collectors.toMap(PassThroughObjectMapper::name, Function.identity()));
        for (var start : passThroughMappers) {
            Set<String> seen = new HashSet<>();
            var current = start;
            while (current != null) {
                if (seen.add(current.name()) == false) {
                    return true;
                }
                current = lookupByName.get(current.supersededBy());
            }
        }
        return false;
```

I think the return type can be a boolean? In the production code the return list is only used to check if an exception should be thrown if the list is not empty?

Reply:
`supersededBy` is a list, I think the code above doesn't take that into account.

The returned list is also included to return a message containing the circular dep, otherwise it may be hard for the end user to understand what's the problem.

Reply:
Not needed any more.